doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,836 @@
+functions.
+Finally, one of the most important steps in the ufunc call is finding the
+correct inner loop for a specific set of input dtypes.  We propose to use
+multiple-dispatching in the future.
+
+.. note::
+
+    At this time this NEP is in a preliminary state. Both internal and
+    external API may be adapted based on user input or implementation needs.
+    The general design principles and choices, while provisional, should not
+    be expected to change dramatically.
+    For example, the *exact* additional information passed into ``UFuncImpl``
+    is likely to evolve.
+
+
+Detailed Description
+--------------------
+
+NEP 41 and NEP 42 layed out the general proposed structure into DType
+classes (similar to the current type numbers) and dtype instances.
+Universal functions have a fairly fixed signature of multiple input arrays
+and certain additional features. But depending on the input DTypes, they
+have to perform a different operation.  In general these operations are
+registered by users, e.g. implementers of a user-defined DTypes.
+
+
+Introduction to General Concepts
+""""""""""""""""""""""""""""""""
+
+Universal functions as described above are much like a Python method
+defined on the DType of the array when considering a ufunc with only
+a single input::
+
+    res = np.positive(arr)
+
+could be implemented (conceptional) as:
+
+    positive_impl = arr.dtype.positive
+    res = positive_impl(arr)
+
+However, unlike methods, ``positive_impl`` is not stored on the dtype itself
+a choice that this NEP does not wish to modify.
+It is rather the implementation for ``np.positive`` for a specific DType.
+Current NumPy partially exposes this "choice of implementation" using
+the ``dtype`` (or more exact ``signature``) attribute in universal functions,
+although these are rarely used:
+
+    np.positive(arr, dtype=np.float64)
+
+forces NumPy to use the ``positive_impl`` written specifically for the Float64
+DType.
+This NEP requires to partially represent this implementation explicitly::
+
+    positive_impl = np.positive.resolve_impl(type(arr.dtype))
+
+Although, this ``positive_impl`` and ``resolve_impl`` is required for certain
+functionality, the following code::
+
+    res = positive_impl(arr)
+
+is not part of this NEP, as it is not central to the proposal.
+
+As mentioned above, in general NumPy universal functions can take many
+inputs, requiring looking up the implementation by considering all of them
+making NumPy ufunc "multi-methods" with respect to the input *DTypes*::
+
+    add_impl = np.add.resolve_impl(type(arr1.dtype), type(arr2.dtype))
+
+This NEP strives to define the minimal set of API necessary to achieve
+the above lists of steps and make it fully extensible for authors of both
+custom ufuncs and DTypes.
+There are two distinct API design decisions as part of this NEP:
+
+1. For the above to work, we must also define::
+
+       np.positive.resolve_impl(MyDateTime)
```

Comment:
Maybe just `MyDType` here and below? Or  in the sentence above, "1. For the above to work for, e.g., a `MyDateTime` DType, we must also define"